Caching: add global cache based on Redis (experimental)

pnax opened the pull request:
## Purpose

Improve our cache implementation to make room to integrate other caching solutions. Provide an optional and experimental global cache layer based on Redis. This can be handy when multiple Zonemaster instances run concurrently (when dealing with batches for instance).

## Context

https://gitlab.rd.nic.fr/zonemaster/zonemaster-engine/-/commit/b96181e460b5824dcd4ac913a3210c458970126a
and
https://gitlab.rd.nic.fr/zonemaster/zonemaster-engine/-/commit/0073398f8a4c47bb6f7f7d8dae8f8a2ddd2694a6

Credits to @blacksponge for the initial proposal

## Changes

* make Cache.pm an interface
* move old Cache.pm logic into a LocalCache module
* new RedisCache module (loaded only if configured in profile.json).

## How to test this PR

1. install Redis and the following perl modules: Redis and Data::MessagePack. For Debian, this can be done with:
     ```
     apt install redis libredis-perl libdata-messagepack-perl
     ```
2. start Redis service
3. modify the profile.json file to specify the Redis server (see `share/profile_additional_properties.json`) :
     ```
     ...
    "cache": {
       "redis": {
          "server": "127.0.0.1:6379",
          "expire": 60
        }
     },
    ...
    ```
4. run the tests or run Zonemaster on a zone to test
5. check that there is data stored in Redis
    ```
    redis-cli KEYS 'ns*'
    ```

matsduf commented on lib/Zonemaster/Engine/Nameserver/Cache/RedisCache.pm:
```diff
@@ -0,0 +1,110 @@
+my $redis;
+our $object_cache = \%Zonemaster::Engine::Nameserver::Cache::object_cache;
+
+my $REDIS_EXPIRE = 3600; # seconds
```
Does that mean that TTL in the DNS record is ignored?

pnax replied:
With this implementation yes. Rethinking about it, it might indeed be a good idea to set an expiry time to `min(TTL, REDIS_EXPIRE)`. I've improved the logic to use the resource record TTL if available.

matsduf commented:
> ## How to test this PR

Don't you have to start the redis service too?

matsduf reviewed (commented):
Documentation of how to run Redis is needed too.

pnax commented:
Todo:
* mark this caching layer as experimental
* improve TTL usage

pnax commented on lib/Zonemaster/Engine/Nameserver/Cache/RedisCache.pm:
```diff
@@ -0,0 +1,118 @@
+    } else {
+        $self->redis->set( $key, '', 'EX', $ttl );
+    }
+}
```
I've updated the logic to take your suggestion in consideration.

pnax commented:
* I've rebased this work on top of `develop`
* I've addressed the remark on negative caching and using the TTL value of the SOA record instead of the minimum field

Please re-review.

pnax commented on lib/Zonemaster/Engine/Nameserver/Cache.pm:
```diff
@@ -5,38 +5,50 @@ use version; our $VERSION = version->declare("v1.0.4");
+        warn "Unknown cache format '$cache', using default 'LocalCache'";
+    }
+    return "LocalCache";
+}
```
The idea was to extend it with other cache format. It looks like I forgot to do the update and add the `RedisCache`. However looking how I did the implementation, I don't use this subroutine at all. I've removed it.

marc-vanderwal commented on lib/Zonemaster/Engine/Nameserver/Cache/RedisCache.pm:
```diff
@@ -29,11 +29,13 @@ if ( $@ ) {
 our $object_cache = \%Zonemaster::Engine::Nameserver::Cache::object_cache;
 
-my $REDIS_EXPIRE = 3600; # seconds
+my $REDIS_EXPIRE = 5; # seconds
```
Why no longer 3600 seconds? Besides, it isn’t sufficiently clear from the variable’s name that it’s merely a default value.

pnax replied:
> Why no longer 3600 seconds?

I recall it being discussed during the F2F. @matsduf wanted a lower value. We agreed on 5 seconds. This is also the value suggested in the additional profile.

pnax commented on lib/Zonemaster/Engine/Nameserver/Cache/RedisCache.pm:
```diff
@@ -0,0 +1,165 @@
+    use Data::MessagePack;
+    use Redis;
+EOS
+);
```
Suggestion applied.

pnax commented on lib/Zonemaster/Engine/Nameserver/Cache/RedisCache.pm:
```diff
@@ -0,0 +1,165 @@
+
+
+if ( $@ ) {
+    die "Cant' use the Redis cache. Make sure the Data::MessagePack and Redis module are installed.\n";
```
fixed

pnax commented:
@tgreenx I should have addressed all your comments/suggestions. Good idea to have a generic entry in the profile for the cache layer.

blacksponge reviewed (dismissed):
I have tested it and have no comment, it seems to be working correctly.

blacksponge commented:
Can you edit the description so that the profile section uses the new `cache` section

matsduf commented:
> I installed redis and the other packages on FreeBSD. Created an updated profile. Ran zonemaster-cli with the updated profile. But the redis cache is empty and rerunning the test does not speed it up. How could I check?
> 
> ```
> # redis-cli KEYS 'ns*'
> (empty array)
> 
> # redis-cli KEYS '*'
> (empty array)
> ```

With the new default, 300 instead of 5 seconds, I get the expected result. Checking "ripe.net" took 50 seconds first time and less than 10 seconds the second time. The `redis-cli` command above also gave a non-empty result.

matsduf commented:
Release testing: A bug was found. Described in #1326 and a solution is provided in #1327. Else it looks fine.
